Readme.md
# mon_sched(1)

  group process supervisor, using a JSON config file, support cron style invoke interval, fork from https://github.com/tj/mon, depends on  https://github.com/udp/json-parser, and the origin `mon` was:

  Super-simple monitoring program.

  `mon` spawned from the needlessly complex
  frustration that tools like [monit](http://mmonit.com/monit/)
  provide, with their awkward DSLs and setup. `mon` is written
  in C, uses less than 400kb of memory, and is incredibly simple
  to set up.

## Usage

```

Usage: mon_sched [options | JSON_CONFIG]

Options:

  -r, <config_json>             run group config
  -v,                           output program version
  -h, 		                output help information
  -s, <pid_file>                check group child process status

```

## Example

  The most simple use of `mon_sched(1)` is to simply keep a command running:

```js
➜  mon_sched git:(master) ✗ ./mon_sched -r example/mon_group.json
mon : mon1 pid 85415
mon : mon2 pid 85416
mon : mon1 sh -c "example/program.sh"
mon : mon2 sh -c "example/program2.sh"
11111
one
22222
two
33333
three
exiting
mon : mon2 exit(1)
mon : mon2 sleep(5)
exiting
mon : mon1 exit(1)
mon : mon1 sleep(10)
mon : mon2 last restart less than one second ago
mon : mon2 30 attempts remaining
mon : mon2 pid 85435
mon : mon2 sh -c "example/program2.sh"
```

  Check group process status

```js
➜  mon_sched git:(master) ✗ ./mon_sched -s mon_sched.pid
mon_group [85592] : alive : uptime 12 seconds
mon1 [-1] : dead
mon2 [85782] : alive : uptime 2 seconds
```

## Config format

```js
{
        "name" : "mon_group",
        "logfile" : "mon_sched.log",
        "pidfile" : "mon_sched.pid",
        "daemon" : false,
        "mon1" : {
                "logfile": "log1.txt",
                "cmd": "example/program.sh",
                "attempts": 20,
		"sleep": 10,
		"on_error" : "",
		"on_restart" : "",
                "cron": "* * * * 2"
        },
        "mon2" : {
                "cmd": "example/program2.sh",
                "logfile": "log2.txt",
                "attempts": 30,
                "sleep": 5,
                "cron": "1 * * * *"
        }
}
```
You may daemonize mon_sched to set group 'daemon' key to true.

and the 'cron' entry format like:

```js
"0,30,45   */3     2,[4-9]     *       *"
# minius   hour    month_day   month   weak_day
# (0-59)   (0-23)  (1-31)      (0-11)  (0-6)
```

## Failure alerts

 `mon_sched(1)` will continue to attempt restarting your program unless the maximum number
 of `attempts` has been exceeded within 60 seconds. Each time a restart is performed
 the `on-restart` command is executed, and when `mon_sched(1)` finally bails the `on-error`
 command is then executed before mon itself exits and gives up.

  For example the following will echo "hey" three times before mon realizes that
  the program is unstable, since it's exiting immediately, thus finally invoking
  `./email.sh`, or any other script you like.

## Logs

  By default `mon_sched(1)` logs to stdio, however when daemonized it will default
  to writing to `/dev/null`.

## Signals

  - __SIGQUIT__ graceful shutdown
  - __SIGTERM__ graceful shutdown

# License

  MIT

# Build Status

  [![Build Status](https://travis-ci.org/lalawue/mon_sched.png)](http://travis-ci.org/lalawue/mon_sched)


History.md

1.3.0 / 2020-01-29
==================

* refactoring as a group child process monitor
* rename to mon_sched and using JSON as config file format

1.2.3 / 2013-12-01
==================

 * improve -S to be position independent of -p
 * improve compatibility with BSD make.

1.2.2 / 2013-06-10 
==================

 * fix waiting for child to exit on graceful shutdown

1.2.1 / 2013-05-31 
==================

 * fix: only invoke error command when specified
 * update commander

1.2.0 / 2012-12-17 
==================

  * add passing of PID to the error scripts

1.1.2 / 2012-12-13 
==================

  * update commander

1.1.1 / 2012-11-12 
==================

  * change SIGKILL to sig

1.1.0 / 2012-11-09 
==================

  * add `--on-error <cmd>`
  * add `--on-restart <cmd>`
  * add `--attempts <max>`
  * add restart limiting / bailing logic
  * change signalling of child to pgid to fix `dash` support

1.0.1 / 2012-11-07 
==================

  * fix child exit, reset signals
  * change graceful_exit to kill() with the same sig

1.0.0 / 2012-11-07 
==================

  * add --on-error <cmd>
  * remove pid log
  * update commander

0.0.3 / 2012-06-15 
==================

  * Refactor with commander

0.0.2 / 2012-05-02 
==================

  * Added: error when an unrecognized flag is passed
  * Added -m, --mon-pidfile. Closes #4
  * Added ms dep



src/mon_sched.c

//
// mon_sched.c
//
// Copyright (c) 2012 TJ Holowaychuk <tj@vision-media.ca>
//
// hacked by lalawue, http://github.com/lalawue

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include <string.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include "ms.h"
#include "config.h"
#include "status.h"

/*
 * Program version.
 */

#define kVERSION "1.3.1"

static mon_t *g_mon;

/*
 * Logger.
 */

#define log(fmt, args...) \
	do { \
		printf("mon : " fmt "\n", ##args); \
		fflush(stdout); \
	} while(0)

/*
 * Output error `msg`.
 */

void
error(char *msg) {
  fprintf(stderr, "Error: %s\n", msg);
  exit(1);
}

/*
 * Return a timestamp in milliseconds.
 */

int64_t
timestamp() {
  struct timeval tv;
  int ret = gettimeofday(&tv, NULL);
  if (-1 == ret) return -1;
  return (int64_t) ((int64_t) tv.tv_sec * 1000 + (int64_t) tv.tv_usec / 1000);
}

/*
 * Write group info to `file`.
 */

void
write_pidfile() {
	mon_dump_status(g_mon);
}

/*
 * Read pid `file`.
 */

pid_t
read_pidfile() {
	return mon_get_pid(g_mon);
}

/*
 * Output status of `pidfile`.
 */

void
show_status_of(const char *pidfile) {
	mon_show_status(pidfile);
}

/*
 * Redirect stdio to `file`.
 */

void
redirect_stdio_to(const char *file) {
	int nullfd = open("/dev/null", O_RDONLY, 0);
	int logfd = nullfd;
	if (file && strncmp(file, "/dev/null", 9) != 0) {
		logfd = open(file, O_WRONLY | O_CREAT | O_APPEND, 0755);  
	}

	if (-1 == logfd) {
		perror("open()");
		exit(1);
	}

	if (-1 == nullfd) {
		perror("open()");
		exit(1);
	}

	dup2(nullfd, 0);
	dup2(logfd, 1);
	dup2(logfd, 2);
}

/*
 * Graceful exit, signal process group.
 */

void
graceful_exit(int sig) {
  int status;
  pid_t pid = getpid();
  log("shutting down");
  log("kill(-%d, %d)", pid, sig);
  kill(-pid, sig);
  log("waiting for exit");
  waitpid(read_pidfile(), &status, 0);
  log("bye :)");
  exit(0);
}

/*
 * Daemonize the program.
 */

void
daemonize() {
  if (fork()) exit(0);

  if (setsid() < 0) {
    perror("setsid()");
    exit(1);
  }
}

/*
 * Invoke the --on-restart command.
 */

void
exec_restart_command(monitor_t *monitor, pid_t pid) {
  char buf[1024] = {0};
  snprintf(buf, 1024, "%s %d", monitor->on_restart, pid);
  log("%s on restart `%s`", monitor->name, buf);
  int status = system(buf);
  if (status) {
	  log("%s exit(%d)", monitor->name, status);
  }
}

/*
 * Invoke the --on-error command.
 */

void
exec_error_command(monitor_t *monitor, pid_t pid) {
  char buf[1024] = {0};
  snprintf(buf, 1024, "%s %d", monitor->on_error, pid);
  log("%s on error `%s`", monitor->name, buf);
  int status = system(buf);
  if (status) {
	  log("%s exit(%d)", monitor->name, status);
  }
}

/*
 * Return the ms since the last restart.
 */

int64_t
ms_since_last_restart(monitor_t *monitor) {
  if (0 == monitor->last_restart_at) return 0;
  int64_t now = timestamp();
  return now - monitor->last_restart_at;
}

/*
 * Check if the maximum restarts within 60 seconds
 * have been exceeded and return 1, 0 otherwise.
 */

int
attempts_exceeded(monitor_t *monitor, int64_t ms) {
	monitor->attempts++;
	monitor->clock -= ms;

	// reset
	if (monitor->clock <= 0) {
		monitor->clock = 60000;
		monitor->attempts = 0;
		return 0;
	}

	// all good
	if (monitor->attempts < monitor->max_attempts) {
		return 0;
	}

	return 1;
}

/*
 * find monitor with pid
 */

static monitor_t*
_monitor_with_pid(pid_t pid)
{
	monitor_t *m = g_mon->monitors;
	while (m) {
		if (pid == m->pid) {
			return m;
		}
		m = m->next_monitor;
	}
	return NULL;
}

/*
 * find monitor in sleep state, increase sleep time, return the close to restart one,
 * if no sleep monitor, return NULL
 */

static monitor_t*
_increase_sleep_monitor(mon_t *mon) {
	int max = 0x7fffffff;
	monitor_t *close_to_restart = NULL;
	monitor_t *m = mon->monitors;
	while (m) {
		if (m->sleepsec > 0) {
			m->sleepsec += 1;
			if (m->max_sleepsec - m->sleepsec < max) {
				max = m->max_sleepsec - m->sleepsec;
				close_to_restart = m;	
			}
		}
		m = m->next_monitor;
	}
	return close_to_restart;
}

/*
 * find cron monitor ready to run, return NULL otherwise
 * set in-active to cro
 */

static monitor_t*
_ready_cron_monitor(mon_t *mon, struct tm *ptm) {
	monitor_t *m = mon->monitors, *ready_m = NULL;
	for ( ; m; m=m->next_monitor) {
		if (!m->cron || m->pid > K_INVALID_MONITOR_PID) {
			continue;
		}
		if (cron_in_timearea(m->cron, ptm)) {
			if (!m->cron->has_running) {
				ready_m = m;
			}
		} else {
			cron_set_has_running(m->cron, false);
		}
	}
	return ready_m;
}

static struct tm*
_current_tm() {
	static struct tm current_tm;
	time_t ti = time(NULL);
	gmtime_r(&ti, &current_tm);
	return &current_tm;
}

static char*
_current_tm_string(struct tm *t) {
	static char buf[32];
	snprintf(buf, 32, "%d/%d/%d %d:%d",
			1900 + t->tm_year, t->tm_mon + 1, t->tm_mday, t->tm_hour, t->tm_min);
	return buf;
}

typedef void (*last_one_callback)();

/*
 * Monitor the given `cmd`.
 */

void
start(monitor_t *monitor, last_one_callback last_callback) {
monitor_exec: {
	if (monitor->cron) {
		// cron skip when not in time area, or has running before
		struct tm *t = _current_tm();
		if (!cron_in_timearea(monitor->cron, t) || monitor->cron->has_running) {
			log("%s cron skip %s", monitor->name, _current_tm_string(t));
			goto monitor_skip;
		}
	}

	pid_t pid = fork();

	if (pid == -1) {
		perror("fork()");
		exit(1);
	} else if (pid == 0)  {		
		signal(SIGTERM, SIG_DFL);
		signal(SIGQUIT, SIG_DFL);
		log("%s sh -c \"%s\"", monitor->name, monitor->cmd);
		execl("/bin/sh", "sh", "-c", monitor->cmd, 0);
		perror("execl()");
		exit(1);
	} else {
		monitor->pid = pid;
		cron_set_has_running(monitor->cron, true);
		log("%s pid %d at %s", monitor->name, pid, _current_tm_string(_current_tm()));

		monitor_skip: {
			if (!last_callback) {
				// callback at last one
				return;
			}
			last_callback();
		}

		// wait cron or sleep monitors
		minitor_wait: {
			int status;
			for (;;) {			
				sleep(1); // always sleep 1 seconds

				monitor = _ready_cron_monitor(g_mon, _current_tm());
				if (monitor) {
					goto monitor_exec;
				}

				pid = waitpid(-1, &status, WNOHANG);				
				if (pid > 0) {
					monitor = _monitor_with_pid(pid);
					monitor->status = status;
					goto monitor_signal;
				} 
				
				monitor = _increase_sleep_monitor(g_mon);
				if (monitor && (monitor->sleepsec > monitor->max_sleepsec)) {
					monitor->sleepsec = 0;
					goto monitor_error;
				}
			}
		}

		monitor_signal: {
			monitor->pid = K_INVALID_MONITOR_PID;
			if (WIFSIGNALED(monitor->status)) {
				log("%s signal(%s)", monitor->name, strsignal(WTERMSIG(monitor->status)));
				log("%s sleep(%d)", monitor->name, monitor->max_sleepsec);
				if (monitor->max_sleepsec > 1) {
					monitor->sleepsec = 1;
					goto minitor_wait;
				} else {
					sleep(1);
					goto monitor_error;
				}
			}
			if (WEXITSTATUS(monitor->status)) {
				log("%s exit(%d)", monitor->name, WEXITSTATUS(monitor->status));
				log("%s sleep(%d)", monitor->name, monitor->max_sleepsec);
				if (monitor->max_sleepsec > 1) {
					monitor->sleepsec = 1;
					goto minitor_wait;
				} else {
					sleep(1);
					goto monitor_error;
				}
			}
		} // monitor_signal
	}

	// restart
	monitor_error: {
		monitor->pid = K_INVALID_MONITOR_PID;
		int64_t ms = ms_since_last_restart(monitor);
		bool exit_normal = (WIFSIGNALED(monitor->status) | WEXITSTATUS(monitor->status)) == 0;
		if (exit_normal || attempts_exceeded(monitor, ms)) {
			if (!exit_normal) {
				char *time = milliseconds_to_long_string(60000 - monitor->clock);
				log("%s %d restarts within %s, bailing", monitor->name, monitor->max_attempts, time);
				if (monitor->on_error) {
					exec_error_command(monitor, pid);
				}
			}
			if (mon_monitor_try_remove(g_mon, monitor)) {
				log("%s bye :)", monitor->name);
			} else {
				mon_monitor_reset(monitor);
			}
			// if no monitors
			if (!g_mon->monitors) {
				log("%s exit, no monitors", g_mon->name);
				exit(2);
			}
		} else {
			log("%s %d attempts remaining", monitor->name, monitor->max_attempts - monitor->attempts);			
			if (monitor->on_restart) {
				exec_restart_command(monitor, pid);
			}
			monitor->last_restart_at = timestamp();
			log("%s last restart %s ago", monitor->name, milliseconds_to_long_string(ms));
		}
		goto monitor_exec;
	}
}}

/*
 * [options] <cmd>
 */

static void
_show_help(char *app_name) {
	printf("Usage:\n");
	printf("%s -r config_json, run with group config\n", app_name);
	printf("%s -v,\t\tshow version\n", app_name);
	printf("%s -h,\t\tshow help\n", app_name);
	printf("%s -s pid_file, show group pid status\n", app_name);
}

int
main(int argc, char *argv[]) {
	if (argc < 3) {
		_show_help(argv[0]);
		return 0;
	}

	if (strcmp(argv[1], "-v") == 0) {
		printf("mon_sched %s\n", kVERSION);
		return 0;
	}

	if (strcmp(argv[1], "-h") == 0) {
		_show_help(argv[0]);
		return 0;
	}

	if (strcmp(argv[1], "-s") == 0) {
		show_status_of(argv[2]);
		return 0;
	}

	if (strcmp(argv[1], "-r") != 0) {
		_show_help(argv[0]);
		return 0;
	}

	// crete global mon instance
	g_mon = mon_create(argv[2]);
	if (g_mon == NULL) {
		return 0;
	} else {
		g_mon->time = time(NULL);
	}
	
	// signals
	signal(SIGTERM, graceful_exit);
	signal(SIGQUIT, graceful_exit);

	// daemonize
  	if (g_mon->daemon) {
    	daemonize();
    	redirect_stdio_to(g_mon->logfile);
  	}

	last_one_callback callback = g_mon->pidfile ? write_pidfile : NULL;

	monitor_t *m = g_mon->monitors;
	while (m) {
	  	start(m, !m->next_monitor ? callback : NULL);
		m = m->next_monitor;
	}

	return 0;
}


src/mon.c

//
// mon.c
//
// Copyright (c) 2012 TJ Holowaychuk <tj@vision-media.ca>
//
// hacked by lalawue, http://github.com/lalawue

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include <string.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include "ms.h"
#include "config.h"

/*
 * Program version.
 */

#define kVERSION "1.3.0"

static mon_t *g_mon;

/*
 * Logger.
 */

#define log(fmt, args...) \
	do { \
		printf("mon : " fmt "\n", ##args); \
		fflush(stdout); \
	} while(0)

/*
 * Output error `msg`.
 */

void
error(char *msg) {
  fprintf(stderr, "Error: %s\n", msg);
  exit(1);
}

/*
 * Check if process of `pid` is alive.
 */

int
alive(pid_t pid) {
  return 0 == kill(pid, 0);
}

/*
 * Return a timestamp in milliseconds.
 */

int64_t
timestamp() {
  struct timeval tv;
  int ret = gettimeofday(&tv, NULL);
  if (-1 == ret) return -1;
  return (int64_t) ((int64_t) tv.tv_sec * 1000 + (int64_t) tv.tv_usec / 1000);
}

/*
 * Write group info to `file`.
 */

void
write_pidfile() {
	if (g_mon->pidfile == NULL) {
		return;
	}
	int fd = open(g_mon->pidfile, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
	if (fd < 0) {
		perror("open()");
	}
	mon_dump_status(g_mon, fd);
	close(fd);
}

/*
 * Read pid `file`.
 */

pid_t
read_pidfile(const char *file) {
  off_t size;
  struct stat s;

  // stat
  if (stat(file, &s) < 0) {
    perror("stat()");
    exit(1);
  }

  size = s.st_size;

  // opens
  int fd = open(file, O_RDONLY, 0);
  if (fd < 0) {
    perror("open()");
    exit(1);
  }

  // read
  char buf[size];
  if (size != read(fd, buf, size)) {
    perror("read()");
    exit(1);
  }

  return atoi(buf);
}

/*
 * Output status of `pidfile`.
 */

void
show_status_of(const char *pidfile) {
  off_t size;
  struct stat s;

  // stat
  if (stat(pidfile, &s) < 0) {
    perror("stat()");
    exit(1);
  }

  size = s.st_size;

  // opens
  int fd = open(pidfile, O_RDONLY, 0);
  if (fd < 0) {
    perror("open()");
    exit(1);
  }

  // read
  char buf[size];
  if (size != read(fd, buf, size)) {
    perror("read()");
    exit(1);
  }

  // uptime
  time_t modified = s.st_mtime;

  struct timeval t;
  gettimeofday(&t, NULL);
  time_t now = t.tv_sec;
  time_t secs = now - modified;

  // status
  pid_t pid = atoi(buf);

  if (alive(pid)) {
    char *str = milliseconds_to_long_string(secs * 1000);
    printf("\e[90m%d\e[0m : \e[32malive\e[0m : uptime %s\e[m\n", pid, str);
    free(str);
  } else {
    printf("\e[90m%d\e[0m : \e[31mdead\e[0m\n", pid);
  }

  close(fd);
}

/*
 * Redirect stdio to `file`.
 */

void
redirect_stdio_to(const char *file) {
	int nullfd = open("/dev/null", O_RDONLY, 0);
	int logfd = nullfd;
	if (file && strncmp(file, "/dev/null", 9) != 0) {
		logfd = open(file, O_WRONLY | O_CREAT | O_APPEND, 0755);  
	}

	if (-1 == logfd) {
		perror("open()");
		exit(1);
	}

	if (-1 == nullfd) {
		perror("open()");
		exit(1);
	}

	dup2(nullfd, 0);
	dup2(logfd, 1);
	dup2(logfd, 2);
}

/*
 * Graceful exit, signal process group.
 */

void
graceful_exit(int sig) {
  int status;
  pid_t pid = getpid();
  log("shutting down");
  log("kill(-%d, %d)", pid, sig);
  kill(-pid, sig);
  log("waiting for exit");
  waitpid(read_pidfile(g_mon->pidfile), &status, 0);
  log("bye :)");
  exit(0);
}

/*
 * Daemonize the program.
 */

void
daemonize() {
  if (fork()) exit(0);

  if (setsid() < 0) {
    perror("setsid()");
    exit(1);
  }
}

/*
 * Invoke the --on-restart command.
 */

void
exec_restart_command(monitor_t *monitor, pid_t pid) {
  char buf[1024] = {0};
  snprintf(buf, 1024, "%s %d", monitor->on_restart, pid);
  log("%s on restart `%s`", monitor->name, buf);
  int status = system(buf);
  if (status) {
	  log("%s exit(%d)", monitor->name, status);
  }
}

/*
 * Invoke the --on-error command.
 */

void
exec_error_command(monitor_t *monitor, pid_t pid) {
  char buf[1024] = {0};
  snprintf(buf, 1024, "%s %d", monitor->on_error, pid);
  log("%s on error `%s`", monitor->name, buf);
  int status = system(buf);
  if (status) {
	  log("%s exit(%d)", monitor->name, status);
  }
}

/*
 * Return the ms since the last restart.
 */

int64_t
ms_since_last_restart(monitor_t *monitor) {
  if (0 == monitor->last_restart_at) return 0;
  int64_t now = timestamp();
  return now - monitor->last_restart_at;
}

/*
 * Check if the maximum restarts within 60 seconds
 * have been exceeded and return 1, 0 otherwise.
 */

int
attempts_exceeded(monitor_t *monitor, int64_t ms) {
  monitor->attempts++;
  monitor->clock -= ms;

  // reset
  if (monitor->clock <= 0) {
    monitor->clock = 60000;
    monitor->attempts = 0;
    return 0;
  }

  // all good
  if (monitor->attempts < monitor->max_attempts) return 0;

  return 1;
}

/*
 * find monitor with pid
 */

static monitor_t*
_monitor_with_pid(pid_t pid)
{
	monitor_t *m = g_mon->monitors;
	while (m) {
		if (pid == m->pid) {
			return m;
		}
		m = m->next_monitor;
	}
	return NULL;
}

/*
 * find monitor in sleep state, increase sleep time, return the close to restart one,
 * also return has_other_child flag
 */

static monitor_t*
_increase_sleep_monitor(mon_t *mon, int *has_other_child) {
	int max = 0x7fffffff;
	monitor_t *close_to_restart = NULL;
	monitor_t *m = mon->monitors;
	while (m) {
		if (m->sleepsec > 0) {
			m->sleepsec += 1;
			if (m->max_sleepsec - m->sleepsec < max) {
				max = m->max_sleepsec - m->sleepsec;
				close_to_restart = m;	
			}
		} else {
			*has_other_child = 1;
		}
		m = m->next_monitor;
	}
	return close_to_restart;
}

typedef void (*last_one_callback)();

/*
 * Monitor the given `cmd`.
 */

void
start(monitor_t *monitor, last_one_callback last_callback) {
monitor_exec: {
	pid_t pid = fork();
	int status;

	if (pid == -1) {
		perror("fork()");
		exit(1);
	} else if (pid == 0)  {
		signal(SIGTERM, SIG_DFL);
		signal(SIGQUIT, SIG_DFL);
		log("%s sh -c \"%s\"", monitor->name, monitor->cmd);
		execl("/bin/sh", "sh", "-c", monitor->cmd, 0);
		perror("execl()");
		exit(1);
	} else {
		log("%s pid %d", monitor->name, pid);
		monitor->pid = pid;

		if (!last_callback) {
			// wait at last one
			return;
		}

		last_callback();

		monitor_sleep: {
			monitor_t *m = NULL;
			int has_other_child = 0;
			while ((m = _increase_sleep_monitor(g_mon, &has_other_child))) {
				sleep(1);
				pid_t tmp_pid = waitpid(-1, &status, WNOHANG);
				if (has_other_child && tmp_pid > 0) {
					pid = tmp_pid;
					monitor = _monitor_with_pid(pid);
					goto monitor_signal;
				} else if (m->sleepsec >= m->max_sleepsec) {
					m->sleepsec = 0;
					monitor = m;
					goto monitor_error;
				}
			}
		}
					
		// wait for exit
		pid = waitpid(-1, &status, 0);
		monitor = _monitor_with_pid(pid);
		if (monitor == NULL) {
			return;
		}

		monitor_signal:
		// signalled		
		if (WIFSIGNALED(status)) {
			log("%s signal(%s)", monitor->name, strsignal(WTERMSIG(status)));
			log("%s sleep(%d)", monitor->name, monitor->max_sleepsec);
			sleep(1);					
			if (monitor->max_sleepsec > 1) {
				monitor->sleepsec = 1;
				goto monitor_sleep;
			} else {
				goto monitor_error;
			}
		}

		// check status
		if (WEXITSTATUS(status)) {
			log("%s exit(%d)", monitor->name, WEXITSTATUS(status));
			log("%s sleep(%d)", monitor->name, monitor->max_sleepsec);
			sleep(1);					
			if (monitor->max_sleepsec > 1) {
				monitor->sleepsec = 1;
				goto monitor_sleep;
			} else {
				goto monitor_error;
			}
		}
	}

	// restart
	monitor_error: {
		if (monitor->on_restart) {
			exec_restart_command(monitor, pid);
		}
		int64_t ms = ms_since_last_restart(monitor);
		monitor->last_restart_at = timestamp();
		log("%s last restart %s ago", monitor->name, milliseconds_to_long_string(ms));
		log("%s %d attempts remaining", monitor->name, monitor->max_attempts - monitor->attempts);
		if (attempts_exceeded(monitor, ms)) {
			char *time = milliseconds_to_long_string(60000 - monitor->clock);
			log("%s %d restarts within %s, bailing", monitor->name, monitor->max_attempts, time);
			if (monitor->on_error) {
				exec_error_command(monitor, pid);
			}
			log("%s bye :)", monitor->name);
			exit(2);
		}
		goto monitor_exec;
	}
}}

/*
 * [options] <cmd>
 */

static void
_show_help(char *app_name) {
	printf("Usage:\n");
	printf("%s JSON_CONFIG\n", app_name);
	printf("%s -v,\t\tshow version\n", app_name);
	printf("%s -h,\t\tshow help\n", app_name);
	printf("%s -s pidfile,\tshow status\n", app_name);
}

int
main(int argc, char *argv[]) {
	if (argc < 2) {
		_show_help(argv[0]);
		return 0;
	}

	if (strcmp(argv[1], "-v") == 0) {
		printf("mon_sched %s\n", kVERSION);
		return 0;
	}

	if (strcmp(argv[1], "-h") == 0) {
		_show_help(argv[0]);
		return 0;
	}

	if (strcmp(argv[1], "-s") == 0) {
		if (argc > 2) {
			char *pidfile = argv[2];
			printf("%s\n", pidfile);
			//FIXME: show pi
		} else {
			_show_help(argv[0]);
		}
		return 0;
	}

	// crete global mon instance
	g_mon = mon_create(argv[1]);
	if (g_mon == NULL) {
		return 0;
	}
	
	// signals
	signal(SIGTERM, graceful_exit);
	signal(SIGQUIT, graceful_exit);

	// daemonize
  	if (g_mon->daemon) {
    	daemonize();
    	redirect_stdio_to(g_mon->logfile);
  	}

	last_one_callback callback = g_mon->pidfile ? write_pidfile : NULL;

	monitor_t *m = g_mon->monitors;
	while (m) {
	  	start(m, !m->next_monitor ? callback : NULL);
		m = m->next_monitor;
	}

	return 0;
}
